subprojects/publish/src/main/java/org/gradle/api/publish/internal/mapping/DefaultDependencyCoordinateResolverFactory.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.publish.internal.mapping;
+
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.DependencyConstraint;
+import org.gradle.api.artifacts.ExternalDependency;
+import org.gradle.api.artifacts.ModuleVersionIdentifier;
+import org.gradle.api.artifacts.ProjectDependency;
+import org.gradle.api.component.SoftwareComponentVariant;
+import org.gradle.api.internal.artifacts.dependencies.DefaultProjectDependencyConstraint;
+import org.gradle.api.internal.artifacts.dependencies.ProjectDependencyInternal;
+import org.gradle.api.internal.artifacts.ivyservice.projectmodule.ProjectDependencyPublicationResolver;
+import org.gradle.api.internal.attributes.AttributeContainerInternal;
+import org.gradle.api.internal.attributes.ImmutableAttributes;
+import org.gradle.api.publish.internal.validation.VariantWarningCollector;
+import org.gradle.api.publish.internal.versionmapping.VariantVersionMappingStrategyInternal;
+import org.gradle.api.publish.internal.versionmapping.VersionMappingStrategyInternal;
+import org.gradle.util.Path;
+
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+
+/**
+ * Default implementation of {@link DependencyCoordinateResolverFactory} that
+ * resolves dependencies using version mapping.
+ */
+public class DefaultDependencyCoordinateResolverFactory implements DependencyCoordinateResolverFactory {
+
+    private final ProjectDependencyPublicationResolver projectDependencyResolver;
+    private final VersionMappingStrategyInternal versionMappingStrategy;
+
+    @Inject
+    public DefaultDependencyCoordinateResolverFactory(
+        ProjectDependencyPublicationResolver projectDependencyResolver,
+        VersionMappingStrategyInternal versionMappingStrategy
+    ) {
+        this.projectDependencyResolver = projectDependencyResolver;
+        this.versionMappingStrategy = versionMappingStrategy;
+    }
+
+    public ComponentDependencyResolver createComponentResolver(SoftwareComponentVariant variant) {
+        return getDependencyResolutionStrategy(variant).componentResolver;
+    }
+
+    @Override
+    public VariantDependencyResolver createVariantResolver(SoftwareComponentVariant variant) {
+        return getDependencyResolutionStrategy(variant).variantResolver;
+    }
```

Comment:
❓ Would it be possible to cache the creation of the `DependencyResolvers` as a field in the factory so as not to have to go through the logic in `getDependencyResolutionStrategy` twice, immediately after the first call, in `MavenComponentParser` lines 160-161?

Reply:
I think caching would be difficult here, given that we'd need to construct a key based on the provided variant. However, I've effectively made the `getDependencyResolutionStrategy` method public (with a different name) so that the maven component parser doesn't need to call this code twice